File: frontend/lib/api/customers.ts
import api from '../api'

export interface CustomerSummary {
  id: string
  client_id: string
  name: string | null
  primary_email: string | null
  primary_phone: string | null
  full_name: string | null
  company_name: string | null
  merged_emails?: string[]
  notes?: string | null
  total_messages?: number
  total_tickets: number
  open_tickets?: number
  total_interactions: number
  last_contacted_at?: string | null
  avg_response_time?: number | null
  sentiment_score?: number | null
  sentiment_label?: string | null
  churn_risk?: string
  avg_satisfaction_score: number | null
  churn_risk_score: number
  last_interaction_at: string | null
  created_at: string | null
}

export interface CustomerTicket {
  id: string
  ticket_id: string
  ticket_number: string
  summary: string
  source: string
  category: string
  priority: number | null
  state: string
  sla_status: string
  resolution_status: string
  assigned_to: string | null
  created_at: string | null
  resolved_at: string | null
}

export interface CustomerInteraction {
  id: string
  interaction_type: string
  interaction_channel: string | null
  summary: string | null
  sentiment_score: number | null
  duration_seconds: number | null
  metadata: Record<string, unknown>
  created_at: string | null
}

export interface CustomerMessage {
  id: string
  customer_id: string | null
  channel: string
  direction: string
  status: string
  sender_id: string | null
  sender_name: string | null
  message_text: string | null
  complaint_id: string | null
  timestamp: string | null
  created_at: string | null
}

export interface CustomerNote {
  id: string
  customer_id: string
  author_email: string
  note_type: string
  content: string
  pinned: boolean
  created_at: string | null
  updated_at: string | null
}

export interface CustomerRelationship {
  id: string
  client_id: string
  parent_customer_id: string
  child_customer_id: string
  relationship_type: string
  role_title: string | null
  is_primary_contact: boolean
  created_at: string | null
}

export interface CustomerDuplicateCandidate {
  customer: CustomerSummary
  confidence_score: number
}

export interface CustomerDetailResponse {
  profile: CustomerSummary
  recent_tickets: CustomerTicket[]
  recent_messages: CustomerMessage[]
  active_tickets: CustomerTicket[]
  interaction_timeline: CustomerInteraction[]
  timeline: CustomerTimelineItem[]
  notes: CustomerNote[]
  relationships: CustomerRelationship[]
  satisfaction_trend: Array<{ week: string; avg_score: number | null }>
  churn_indicators: Record<string, unknown>
  sentiment?: {
    score: number | null
    label: string
    sample_size: number
  }
  insights?: string[]
  stats: {
    total_messages?: number
    total_tickets: number
    open_tickets?: number
    total_interactions: number
    last_contacted_at?: string | null
    avg_response_time?: number | null
    avg_satisfaction: number | null
    churn_risk: number
    lifetime_value: number
  }
}

export interface CustomerTimelineItem {
  id: string
  type: 'message' | 'ticket' | 'action'
  title: string
  body: string
  channel?: string
  direction?: string
  status?: string
  timestamp: string | null
  sort_at?: string | null
  data: Record<string, unknown>
}

export interface Customer360Response {
  identity: {
    id: string
    client_id: string
    name: string | null
    primary_email: string | null
    merged_emails: string[]
    tags: string[]
    notes: string | null
    created_at: string | null
    updated_at: string | null
  }
  metrics: {
    total_messages: number
    total_tickets: number
    open_tickets: number
    last_contacted_at: string | null
    avg_response_time: number | null
  }
  sentiment: {
    score: number | null
    label: string
    sample_size: number
  }
  churn_risk: 'low' | 'medium' | 'high'
  recent_messages: CustomerMessage[]
  recent_tickets: CustomerTicket[]
  active_tickets: CustomerTicket[]
  timeline: CustomerTimelineItem[]
  insights: string[]
}

export const customersAPI = {
  list: async (params?: { search?: string; skip?: number; limit?: number }) => {
    const response = await api.get('/api/v1/customers', { params })
    return response.data as { total: number; items: CustomerSummary[] }
  },

  getById: async (customerId: string) => {
    const response = await api.get(`/api/v1/customers/${customerId}`)
    return response.data as CustomerDetailResponse
  },

  get360: async (customerId: string) => {
    const response = await api.get(`/api/v1/customers/${customerId}/360`)
    return response.data as Customer360Response
  },

  getDuplicates: async (customerId: string) => {
    const response = await api.get(`/api/v1/customers/${customerId}/duplicates`)
    return response.data as { customer_id: string; potential_duplicates: CustomerDuplicateCandidate[] }
  },

  addNote: async (customerId: string, payload: { content: string; note_type?: string; pinned?: boolean }) => {
    const response = await api.post(`/api/v1/customers/${customerId}/notes`, payload)
    return response.data as { success: boolean; note: CustomerNote }
  },

  update: async (customerId: string, payload: { name?: string; notes?: string; tags?: string[] }) => {
    const response = await api.patch(`/api/v1/customers/${customerId}`, payload)
    return response.data as { success: boolean; customer: CustomerSummary }
  },
}

File: app/api/v1/customers.py
import uuid
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.db.models import Customer, CustomerNote, CustomerRelationship
from app.db.session import get_db
from app.dependencies.auth import require_api_key
from app.middleware.feature_gate import ensure_feature_access
from app.services.customer_deduplication import CustomerDeduplicator
from app.services.customer_profile import (
    CustomerProfileService,
    serialize_customer,
    serialize_customer_interaction,
    serialize_customer_message,
    serialize_customer_note,
    serialize_customer_relationship,
    serialize_customer_ticket,
)

router = APIRouter(prefix="/api/v1/customers", tags=["customers-v1"])


def _parse_customer_id(customer_id: str) -> uuid.UUID:
    try:
        return uuid.UUID(str(customer_id))
    except (TypeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail="Invalid customer id") from exc


def _actor_email(client) -> str:
    return f"client-{str(client.id)[:8]}@system.local"


def _get_customer_or_404(db: Session, customer_id: uuid.UUID, client_id) -> Customer:
    customer = db.query(Customer).filter(Customer.id == customer_id, Customer.client_id == client_id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    return customer


class MergeCustomersRequest(BaseModel):
    master_customer_id: str
    duplicate_customer_id: str
    merged_by: Optional[str] = None


class CustomerNoteRequest(BaseModel):
    content: str = Field(..., min_length=1)
    note_type: str = "general"
    pinned: bool = False
    author_email: Optional[str] = None


class CustomerRelationshipRequest(BaseModel):
    child_customer_id: str
    relationship_type: str = Field(..., min_length=1)
    role_title: Optional[str] = None
    is_primary_contact: bool = False


class UpdateCustomerRequest(BaseModel):
    name: Optional[str] = None
    notes: Optional[str] = None
    tags: Optional[list[str]] = None


@router.get("")
def list_customers(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    search: Optional[str] = Query(default=None),
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    query = db.query(Customer).filter(Customer.client_id == current_client.id, Customer.is_master == True)
    if search:
        pattern = f"%{search.strip()}%"
        query = query.filter(
            or_(
                Customer.name.ilike(pattern),
                Customer.full_name.ilike(pattern),
                Customer.primary_email.ilike(pattern),
                Customer.company_name.ilike(pattern),
                Customer.primary_phone.ilike(pattern),
            )
        )

    total = query.count()
    customers = query.order_by(Customer.last_contacted_at.desc(), Customer.updated_at.desc()).offset(skip).limit(limit).all()
    return {"total": total, "items": [serialize_customer(customer) for customer in customers]}


@router.post("/merge")
def merge_customers(
    request: MergeCustomersRequest,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    master = _get_customer_or_404(db, _parse_customer_id(request.master_customer_id), current_client.id)
    duplicate = _get_customer_or_404(db, _parse_customer_id(request.duplicate_customer_id), current_client.id)
    try:
        merged = CustomerDeduplicator(db).merge_customers(
            master_id=str(master.id),
            duplicate_id=str(duplicate.id),
            merged_by=(request.merged_by or current_client.name or _actor_email(current_client)),
            merge_strategy="manual",
            confidence_score=1.0,
            commit=True,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {"success": True, "master_customer": serialize_customer(merged)}


@router.get("/{customer_id}")
def get_customer_360(
    customer_id: str,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    parsed_customer_id = _parse_customer_id(customer_id)
    customer = _get_customer_or_404(db, parsed_customer_id, current_client.id)

    service = CustomerProfileService(db)
    data = service.get_customer_360(str(customer.id))

    profile = data["profile"]
    if str(profile.client_id) != str(current_client.id):
        raise HTTPException(status_code=403, detail="Access denied")

    return {
        "profile": serialize_customer(profile),
        "recent_tickets": [serialize_customer_ticket(ticket) for ticket in data["recent_tickets"]],
        "recent_messages": [serialize_customer_message(message) for message in data.get("recent_messages", [])],
        "active_tickets": [serialize_customer_ticket(ticket) for ticket in data.get("active_tickets", [])],
        "interaction_timeline": [serialize_customer_interaction(item) for item in data["interaction_timeline"]],
        "timeline": data.get("timeline", []),
        "notes": [serialize_customer_note(item) for item in data["notes"]],
        "relationships": [serialize_customer_relationship(item) for item in data["relationships"]],
        "satisfaction_trend": data["satisfaction_trend"],
        "churn_indicators": data["churn_indicators"],
        "sentiment": data.get("sentiment"),
        "insights": data.get("insights", []),
        "stats": data["stats"],
    }


@router.get("/{customer_id}/360")
def get_customer_360_snapshot(
    customer_id: str,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    snapshot = CustomerProfileService(db).get_customer_360_snapshot(str(customer.id))
    if snapshot["identity"]["client_id"] != str(current_client.id):
        raise HTTPException(status_code=403, detail="Access denied")
    return snapshot


@router.patch("/{customer_id}")
def update_customer(
    customer_id: str,
    request: UpdateCustomerRequest,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    update_data = request.model_dump(exclude_unset=True)
    if "name" in update_data:
        normalized_name = (request.name or "").strip() or None
        customer.name = normalized_name
        customer.full_name = normalized_name
    if "notes" in update_data:
        customer.notes = (request.notes or "").strip() or None
    if "tags" in update_data and request.tags is not None:
        customer.tags = [tag for tag in dict.fromkeys((tag or "").strip() for tag in request.tags) if tag]
    CustomerProfileService(db).refresh_customer_metrics(customer, commit=False)
    db.commit()
    db.refresh(customer)
    return {"success": True, "customer": serialize_customer(customer)}


@router.get("/{customer_id}/duplicates")
def find_duplicates(
    customer_id: str,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    duplicates = CustomerDeduplicator(db).find_duplicates(customer, limit=10)
    return {
        "customer_id": str(customer.id),
        "potential_duplicates": [
            {
                "customer": serialize_customer(duplicate),
                "confidence_score": confidence,
            }
            for duplicate, confidence in duplicates
        ],
    }


@router.get("/{customer_id}/notes")
def list_customer_notes(
    customer_id: str,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    notes = (
        db.query(CustomerNote)
        .filter(CustomerNote.customer_id == customer.id)
        .order_by(CustomerNote.pinned.desc(), CustomerNote.created_at.desc())
        .all()
    )
    return {"items": [serialize_customer_note(note) for note in notes]}


@router.post("/{customer_id}/notes")
def create_customer_note(
    customer_id: str,
    request: CustomerNoteRequest,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    note = CustomerNote(
        customer_id=customer.id,
        author_email=(request.author_email or _actor_email(current_client)).strip(),
        note_type=request.note_type.strip() or "general",
        content=request.content.strip(),
        pinned=request.pinned,
    )
    db.add(note)
    db.commit()
    db.refresh(note)
    return {"success": True, "note": serialize_customer_note(note)}


@router.get("/{customer_id}/relationships")
def list_customer_relationships(
    customer_id: str,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    relationships = (
        db.query(CustomerRelationship)
        .filter(
            or_(
                CustomerRelationship.parent_customer_id == customer.id,
                CustomerRelationship.child_customer_id == customer.id,
            )
        )
        .order_by(CustomerRelationship.created_at.desc())
        .all()
    )
    return {"items": [serialize_customer_relationship(item) for item in relationships]}


@router.post("/{customer_id}/relationships")
def create_customer_relationship(
    customer_id: str,
    request: CustomerRelationshipRequest,
    db: Session = Depends(get_db),
    current_client=Depends(require_api_key),
):
    parent_customer = _get_customer_or_404(db, _parse_customer_id(customer_id), current_client.id)
    child_customer = _get_customer_or_404(db, _parse_customer_id(request.child_customer_id), current_client.id)
    if parent_customer.id == child_customer.id:
        raise HTTPException(status_code=400, detail="Cannot relate a customer to itself")

    existing = (
        db.query(CustomerRelationship)
        .filter(
            CustomerRelationship.parent_customer_id == parent_customer.id,
            CustomerRelationship.child_customer_id == child_customer.id,
        )
        .first()
    )
    if existing:
        existing.relationship_type = request.relationship_type.strip()
        existing.role_title = request.role_title
        existing.is_primary_contact = request.is_primary_contact
        db.commit()
        db.refresh(existing)
        return {"success": True, "relationship": serialize_customer_relationship(existing)}

    relationship = CustomerRelationship(
        client_id=current_client.id,
        parent_customer_id=parent_customer.id,
        child_customer_id=child_customer.id,
        relationship_type=request.relationship_type.strip(),
        role_title=request.role_title,
        is_primary_contact=request.is_primary_contact,
    )
    db.add(relationship)
    db.commit()
    db.refresh(relationship)
    return {"success": True, "relationship": serialize_customer_relationship(relationship)}
